tools/tpm2_duplicate.c

```diff
@@ -0,0 +1,444 @@
+                " missing option.");
+        result = false;
+    }
+
+    return result;
+}
+
+static bool load_object(ESYS_CONTEXT *ectx, const char *arg, tpm2_loaded_object * obj)
+{
+    bool result = true;
+    tpm2_object_load_rc rc = tpm2_util_object_load(ectx, arg, obj);
+
+    if (rc != olrc_error) {
+        if (!obj->tr_handle) {
+            result = tpm2_util_sys_handle_to_esys_handle(ectx, obj->handle, &obj->tr_handle);
+        }
+    }
+    return result;
+}
+
+static bool set_key_algorithm(TPMI_ALG_PUBLIC alg, TPMT_SYM_DEF_OBJECT * obj)
+{
+    bool result = true;
+    switch (alg) {
+
+    case TPM2_ALG_AES :
+        obj->algorithm = TPM2_ALG_AES;
+        obj->keyBits.aes = 128;
+        obj->mode.aes = TPM2_ALG_CFB;
+        break;
+
+    case TPM2_ALG_NULL :
+        obj->algorithm = TPM2_ALG_NULL;
+        break;
+
+    default:
+        LOG_ERR("The algorithm type input(%4.4x) is not supported!", alg);
+        result = false;
+        break;
+    }
+
+    return result;
+}
+
+static bool load_sym_key(TPM2B_DATA * in_key)
+{
+    FILE *f;
+    unsigned long file_size = 0;
+    bool result;
+
+    f = fopen(ctx.sym_key_in, "r");
+    if (!f) {
+        LOG_ERR("Could not open file \"%s\", error: %s",
+                ctx.sym_key_in, strerror(errno));
+        return false;
+    }
+    /*
+    * Get the file size and validate that it is the proper AES keysize.
+    */
+    result = files_get_file_size(f, &file_size, ctx.sym_key_in);
+    if (!result) {
+        return false;
+    }
+
+    if(file_size != 16) {
+        LOG_ERR("Invalid AES key size, got %lu bytes, expected 16",
+                file_size);
+        return false;
+    }
+
+    in_key->size = file_size;
+    result = files_read_bytes(f, in_key->buffer, in_key->size);
+    fclose(f);
+    if(!result) {
+        return false;
+    }
+    return true;
+}
+
+int tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {
+
+    UNUSED(flags);
+
+    int rc = 1;
+    bool result;
+    TPMT_SYM_DEF_OBJECT sym_alg;
+    TPM2B_DATA in_key;
+    TPM2B_DATA* out_key;
+    TPM2B_PRIVATE* duplicate;
+    TPM2B_ENCRYPTED_SECRET* outSymSeed;
+
+    result = check_options();
+    if (!result) {
+        goto out;
+    }
+
+    result = load_object(ectx, ctx.object_arg, &ctx.object_context);
+    if(!result) {
+        goto out;
+    }
+
+    result = load_object(ectx, ctx.new_parent_object_arg, &ctx.new_parent_object_context);
+    if(!result) {
+        goto out;
+    }
+
+    if (ctx.flags.p) {
+        result = tpm2_auth_util_from_optarg(ectx, ctx.object_auth_str,
+            &ctx.auth.session_data, &ctx.auth.session);
+        if (!result) {
+            LOG_ERR("Invalid authorization, got\"%s\"", ctx.object_auth_str);
+            goto out;
+        }
+    }
+
+    result = set_key_algorithm(ctx.key_type, &sym_alg);
+    if(!result) {
+        goto out;
+    }
+
+    if(ctx.flags.k) {
+        result = load_sym_key(&in_key);
```

Comment:
Why not using `files_load_bytes_from_path` from `lib/files.h` with `size = 16` ?

Reply:
I did originally. Then i looked into tpm2_openssl.c:load_private_AES_from_file() & I decided to follow that general pattern instead. I can change it back if necessary

Reply:
Yeah use that and just drop the routine, then you don't have to deal with the comments. That tpm2_openssl.c file probably should use something like that too.